/*
 * Copyright Â© 2023- Gimlet Labs, Inc.
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Gimlet Labs, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Gimlet Labs, Inc. and its suppliers and
 * may be covered by U.S. and Foreign Patents, patents in process,
 * and are protected by trade secret or copyright law. Dissemination
 * of this information or reproduction of this material is strictly
 * forbidden unless prior written permission is obtained from
 * Gimlet Labs, Inc.
 *
 * SPDX-License-Identifier: Proprietary
 */

syntax = "proto3";

package gml.internal.api.core.v1;

option go_package = "gimletlabs.ai/gimlet/src/api/corepb/v1;corepb";

import "third_party/github.com/gogo/protobuf/gogoproto/gogo.proto";
import "src/common/typespb/uuid.proto";
import "google/protobuf/timestamp.proto";

/**
 * This file contains all the messages that are shared between the control plane and edge.
 */

/****************************************************************************
 * Messages that are used to monitor health status, routing, etc.
 *****************************************************************************/

// EdgeHeartbeat is the message sent from the edge device to let the control plane know it's
// connected and responsive.
message EdgeHeartbeat {
  // The sequence ID of this heartbeat. This needs to be monotonically increasing.
  int64 seq_id = 1 [ (gogoproto.customname) = "SeqID" ];
}

/****************************************************************************
 * Wrapper messages for each channel on the NATS bus from the edge.
 * They are named EdgeCP[*] for messages originating from the edge.
 *****************************************************************************/

// EdgeCPMetadata has basic routing information for shared bus.
message EdgeCPMetadata {
  // The device ID where the message is originating from.
  gml.types.UUID device_id = 1 [ (gogoproto.customname) = "DeviceID" ];
  // The timestamp when the message was received.
  google.protobuf.Timestamp recv_timestamp = 2;
}

// EdgeCPStatus has status message and is on the .Status bus.
message EdgeCPStatus {
  // Associated metadata.
  EdgeCPMetadata metadata = 1;

  oneof msg_oneof {
    EdgeHeartbeat heartbeat = 1000;
  }
}
